usr/src/tools/scripts/xref.1
.\" ident	"@(#)xref.1	1.4	06/02/12 SMI"
.\" " CDDL HEADER START
.\" "
.\" " The contents of this file are subject to the terms of the
.\" " Common Development and Distribution License, Version 1.0 only
.\" " (the "License").  You may not use this file except in compliance
.\" " with the License.
.\" "
.\" " You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
.\" " or http://www.opensolaris.org/os/licensing.
.\" " See the License for the specific language governing permissions
.\" " and limitations under the License.
.\" "
.\" " When distributing Covered Code, include this CDDL HEADER in each
.\" " file and include the License file at usr/src/OPENSOLARIS.LICENSE.
.\" " If applicable, add the following below this CDDL HEADER, with the
.\" " fields enclosed by brackets "[]" replaced with your own identifying
.\" " information: Portions Copyright [yyyy] [name of copyright owner]
.\" "
.\" " CDDL HEADER END
.\" "
.\" "Copyright 2006 Sun Microsystems, Inc.  All rights reserved."
.\" "Use is subject to license terms."
.TH xref 1 "10 Feb 2006"
.SH NAME
.I xref
\- build and maintain source cross-references
.SH SYNOPSIS
\fBxref [-cfp] [-m <makefile>] [-x cscope|ctags|etags[,...]] [<subtree> ...]\fP
.LP
.SH DESCRIPTION
.IX "OS-Net build tools" "xref" "" "\fBxref\fP"
.LP
.I xref
is a simple tool for building and maintaining source cross-references for
subtrees of a workspace.  By default, xref builds cscope-fast, ctags, and
etags (if available; see NOTES) source cross-references for all of the
eligible files underneath each named subtree, plus any additional eligible
files reported by flg.flp(1).  File eligibility is determined by pattern
matching -- by default, files matching the following patterns are
eligible:

	*.[Ccshlxy] Makefile* *.il* *.cc *.adb llib-*

However, this default behavior can be changed, either through command-line
options or by setting the cross-reference customization macros (see
\fICUSTOMIZATION MACROS\fP) in the Makefile rooted at each specified
subtree.
.LP
Any number of subtrees can be specified; if no subtrees are specified,
then the current working directory is assumed.  Relative subtrees are
assumed to be relative to the current working directory.
.SH OPTIONS
.TP 10
.B -c
Clobber (rather than build) cross-references.
.TP 10
.B -f
Do not use flg.flp(1) to locate additional related files.  This option is
ignored if \fB-c\fP is specified.
.TP
.B -m \fImakefile\fP
Use the customization macros specified in \fImakefile\fP when building the
cross-reference, rather than those in Makefile.  If no Makefile exists
or no customizations are desired, then /dev/null may be used.
.TP 10
.B -p
Remove any proto-area include directories from the include path list when
building the cross-reference.  This is chiefly useful when the built
cross-reference will be shared with others that either cannot or should
not access the proto areas associated with the workspace used to construct
the cross-reference. This option is ignored if \fB-c\fP is specified.
.TP 10
.B -x cscope|tags|etags
Operate on only the specified cross-references.  More than one
cross-reference can be operated on by specifying a comma-separated list
(see \fIEXAMPLES\fP).  By default, all supported cross-references are
operated on.
.SH CUSTOMIZATION MACROS
The xref tool uses a private Makefile, xref.mk, to perform the actual
construction and maintenance of each cross-reference.  While the logic in
xref.mk is sufficient for most situations, some subtrees of a workspace
may require additional fine-tuning (e.g., to exclude machine-generated
source files, or to include source files with unusual filenames).  To
accomodate this, xref.mk itself \fBincludes\fP the the file named
\fIMakefile\fP (or the filename specified via -m) at the root of each
specified subtree.  This Makefile in turn may customize the
cross-reference by setting any of the following macros:
.LP
.B XRDIRS
.RS 4
The list of directories to build the cross-reference over; defaults to
`.'.  Directories containing source files most likely to be of interest to
those using the cross-reference should be listed earlier.
.RE
.LP
.B XRPRUNE
.RS 4
The list of directories to prune out.
.RE
.LP
.B XRADD
.RS 4
The list of additional filename patterns to include.
.RE
.LP
.B XRDEL
.RS 4
The list of additional filename patterns to exclude.
.RE
.LP
XRINCDIRS
.RS 4
The list of additional include paths, in "foo bar" format.
.RE
.LP
.B XRINCS
.RS 4
The list of additional include paths, in "-Ifoo -Ibar" format.
.RE
.SH EXAMPLES
Build cscope-fast, ctags, and etags cross-references for the subtree
rooted at the current working directory:
.LP
$ xref
.LP
Build cscope-fast, ctags, and etags cross-references for $SRC/cmd/agents:
.LP
$ xref $SRC/cmd/agents
.LP
Build just the cscope-fast cross-reference for $SRC, and do not use
flg.flp(1) (since there are no additional source files to find):
.LP
$ xref -f -x cscope $SRC
.LP
Build just the standard cscope cross-reference for $SRC/uts:
.LP
$ CSCOPE=cscope CSFLAGS=-b xref -x cscope $SRC/uts
.LP
Build a cscope database containing just the ZFS source:
.LP
.nf
$ cd $SRC
$ cat > /tmp/zfs.mk
XRDIRS=uts/common/fs/zfs common/zfs lib/libzfs cmd/zfs cmd/zpool
.fi
$ xref -f -x cscope -m /tmp/zfs.mk
.LP
.SH ENVIRONMENT VARIABLES
Because xref uses xref.mk to carry out the requested operations and
invokes make(1) with the -e option, the value of any macros used by
xref.mk can be overridden through environment variables.  This is 
chiefly useful for customizing the programs used to build the cross
references; see \fIEXAMPLES\fP.
.LP
.SH SEE ALSO
.LP
.IR make(1),
.IR bringover(1),
.IR flg.flp(1)
.SH NOTES
The etags utility is no longer bundled with the Sun Studio, which
means it may not be available on all developer systems.  If xref cannot
locate the etags utility and the -x option has not been specified, then
xref will silently not build etags.
.LP
The ETAGS environment variable can be used to specify an alternate
path to the etags utility.  For instance, to use the one packaged with
Sun Freeware, set ETAGS to /opt/sfw/bin/etags prior to invoking xref.


usr/src/tools/scripts/xref.sh
#! /bin/ksh -p
#
# CDDL HEADER START
#
# The contents of this file are subject to the terms of the
# Common Development and Distribution License, Version 1.0 only
# (the "License").  You may not use this file except in compliance
# with the License.
#
# You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE
# or http://www.opensolaris.org/os/licensing.
# See the License for the specific language governing permissions
# and limitations under the License.
#
# When distributing Covered Code, include this CDDL HEADER in each
# file and include the License file at usr/src/OPENSOLARIS.LICENSE.
# If applicable, add the following below this CDDL HEADER, with the
# fields enclosed by brackets "[]" replaced with your own identifying
# information: Portions Copyright [yyyy] [name of copyright owner]
#
# CDDL HEADER END
#
#
# Copyright 2002-2003 Sun Microsystems, Inc.  All rights reserved.
# Use is subject to license terms.
#
# ident	"@(#)xref.sh	1.4	05/06/08 SMI"
#
# xref: build and maintain source cross-reference databases.
#

PATH=/usr/bin:/usr/ccs/bin:${BUILD_TOOLS:-/opt}/teamware/bin export PATH
PROG=`basename $0`
XREFMK=`dirname $0`/xref.mk
XRMAKEFILE=Makefile export XRMAKEFILE

#
# The CSCOPEOPTIONS variable can cause problems if it's set in the environment
# when using cscope; remove it.
#
unset CSCOPEOPTIONS

#
# The CDPATH variable causes ksh's `cd' builtin to emit messages to stdout
# under certain circumstances, which can really screw things up; unset it.
#
unset CDPATH

#
# Print the provided failure message and exit with an error.
#
fail()
{
        echo $PROG: $@ > /dev/stderr
        exit 1
}

#
# Print the provided warning message.
#
warn()
{
        echo $PROG: warning: $@ > /dev/stderr
}

#
# Print the provided informational message.
#
info()
{
        echo $PROG: $@
}

#
# Print the provided informational message, and the current value of $SECONDS
# in a user-friendly format.
#
timeinfo()
{
	typeset -Z2 sec
	typeset min seconds

	((seconds = SECONDS))
	((min = seconds / 60))
	((sec = seconds % 60))

	info "$1 in ${min}m${sec}s"
}

#
# If $CODEMGR_WS isn't set, then attempt to glean it from the workspace
# command before giving up.
#
if [ -z "$CODEMGR_WS" ]; then
	if whence workspace > /dev/null; then
		#
		# Since ws(1) hasn't been run, set up SRC and MACH too.
		# Note that other environment variables, such as
		# ENVCPPFLAGS*, can also affect the resulting
		# cross-reference, but we assume that if the developer
		# really cared, he would've ws'd first.
		#
		CODEMGR_WS=`workspace name` export CODEMGR_WS
		SRC=$CODEMGR_WS/usr/src export SRC
		MACH=`uname -p` export MACH
	else
		fail "No active workspace; run \"ws <workspace_name>\""
	fi
fi

[ -d "$CODEMGR_WS" ] || fail "\$CODEMGR_WS ($CODEMGR_WS) is not a directory"
[ -f $XREFMK ] || fail "cannot locate xref.mk"

clobber=
noflg=
xrefs=

while getopts cfm:px: flag; do
	case $flag in
	c)
		clobber=y
		;;
	f)
		noflg=y
		;;
	m)
		XRMAKEFILE=$OPTARG
		;;
	p)
		#
		# The ENVCPPFLAGS* environment variables contain the include
		# paths to our proto areas; clear 'em so that they don't end
		# up in CPPFLAGS, and thus don't end up in XRINCS in xref.mk.
		#
		ENVCPPFLAGS1=
		ENVCPPFLAGS2=
		ENVCPPFLAGS3=
		ENVCPPFLAGS4=
		;;
 	x)
		xrefs=$OPTARG
		;;
	\?)
		echo "usage: $PROG [-cfp] [-m <makefile>]"\
		     "[-x cscope|ctags|etags[,...]] [<subtree> ...]"\
		      > /dev/stderr
		exit 1
		;;
	esac
done

shift $((OPTIND - 1))

#
# Get the list of directories before we reset $@.
#
dirs=$@
[ -z "$dirs" ] && dirs=.

#
# Get the canonical path to the workspace.  This allows xref to work
# even in the presence of lofs(7FS).
#
cd $CODEMGR_WS
CODEMGR_WS=`/bin/pwd`
cd - > /dev/null

#
# Process the xref format list.  For convenience, support common synonyms
# for the xref formats.
#
if [ -z "$xrefs" ]; then
	#
	# Disable etags if we can't find it.
	#
	xrefs="cscope ctags"
	make -e -f $XREFMK xref.etags.check 2>/dev/null 1>&2 && \
	    xrefs="$xrefs etags"
else
	oldifs=$IFS
	IFS=,
	set -- $xrefs
	IFS=$oldifs

	xrefs=
	for xref; do
		case $xref in
		cscope|cscope.out)
			xrefs="$xrefs cscope"
			;;
		ctags|tags)
			xrefs="$xrefs ctags"
			;;
		etags|TAGS)
			xrefs="$xrefs etags"
			;;
		*)
			warn "ignoring unknown cross-reference \"$xref\""
			;;
 		esac
 	done

	[ -z "$xrefs" ] && fail "no known cross-reference formats specified"
fi

#
# Process the requested list of directories.
#
for dir in $dirs; do
	if [ ! -d $dir ]; then
		warn "directory \"$dir\" does not exist; skipping"
		continue
	fi

	#
	# NOTE: we cannot use $PWD because it will mislead in the presence
	# of lofs(7FS).
	#
	cd $dir || fail "cannot change to directory $dir"
	pwd=`/bin/pwd`
	reldir=${pwd##${CODEMGR_WS}/}
	if [ "$reldir" = "$pwd" ]; then
		warn "directory \"$pwd\" is not beneath \$CODEMGR_WS; skipping"
		cd - > /dev/null
		continue
	fi

	#
	# If we're building cross-references, then run `xref.clean' first
	# to purge any crud that may be lying around from previous aborted runs.
	#
	if [ -z "$clobber" ]; then
		make -e -f $XREFMK xref.clean > /dev/null
	fi

	#
	# Find flg-related source files, if requested.
	#
	if [ -z "$noflg" -a -z "$clobber" ]; then
		SECONDS=0
    		info "$reldir: finding flg-related source files"
		make -e -f $XREFMK xref.flg > /dev/null
		if [ $? -ne 0 ]; then
			warn "$reldir: unable to find flg-related source files"
		else
			nfiles=`wc -l < xref.flg`
			if [ "$nfiles" -eq 1 ]; then
				msg="found 1 flg-related source file"
			else
				msg="found $nfiles flg-related source files"
			fi
			timeinfo "$reldir: $msg"
		fi
	fi

	#
	# Build or clobber all of the requested cross-references.
	#
	for xref in $xrefs; do
		if [ -n "$clobber" ]; then
			info "$reldir: clobbering $xref cross-reference"
			make -e -f $XREFMK xref.${xref}.clobber > /dev/null ||
 			    warn "$reldir: cannot clobber $xref cross-reference"
			continue
		fi

		SECONDS=0
		info "$reldir: building $xref cross-reference"
		make -e -f $XREFMK xref.${xref} > /dev/null ||
		    fail "$reldir: cannot build $xref cross-reference"
		timeinfo "$reldir: built $xref cross-reference"
 	done

	make -e -f $XREFMK xref.clean > /dev/null ||
	    warn "$reldir: cannot clean up temporary files"
	cd - > /dev/null
done
exit 0
